.dotnet/examples/Chat/Example04_FunctionCallingStreaming.cs

```diff
@@ -12,6 +15,122 @@ public partial class ChatExamples
 {
     // See Example03_FunctionCalling.cs for the tool and function definitions.
 
+    #region
+    public class StreamingChatToolCallsBuilder
+    {
+        private readonly Dictionary<int, string> _indexToToolCallId = [];
+        private readonly Dictionary<int, string> _indexToFunctionName = [];
+        private readonly Dictionary<int, SequenceBuilder<byte>> _indexToFunctionArguments = [];
+
+        public void Append(StreamingChatToolCallUpdate toolCallUpdate)
+        {
+            // Keep track of which tool call ID belongs to this update index.
+            if (toolCallUpdate.Id != null)
+            {
+                _indexToToolCallId[toolCallUpdate.Index] = toolCallUpdate.Id;
+            }
+
+            // Keep track of which function name belongs to this update index.
+            if (toolCallUpdate.FunctionName != null)
+            {
+                _indexToFunctionName[toolCallUpdate.Index] = toolCallUpdate.FunctionName;
+            }
+
+            // Keep track of which function arguments belong to this update index,
+            // and accumulate the arguments as new updates arrive.
+            if (toolCallUpdate.FunctionArgumentsUpdate != null && !toolCallUpdate.FunctionArgumentsUpdate.ToMemory().IsEmpty)
+            {
+                if (!_indexToFunctionArguments.TryGetValue(toolCallUpdate.Index, out SequenceBuilder<byte> argumentsBuilder))
+                {
+                    argumentsBuilder = new SequenceBuilder<byte>();
+                    _indexToFunctionArguments[toolCallUpdate.Index] = argumentsBuilder;
+                }
+
+                argumentsBuilder.Append(toolCallUpdate.FunctionArgumentsUpdate);
+            }
+        }
+
+        public IReadOnlyList<ChatToolCall> Build()
+        {
+            List<ChatToolCall> toolCalls = [];
+
+            foreach ((int index, string toolCallId) in _indexToToolCallId)
+            {
+                ChatToolCall toolCall = ChatToolCall.CreateFunctionToolCall(
+                    toolCallId: toolCallId,
+                    functionName: _indexToFunctionName[index],
+                    functionArguments: BinaryData.FromBytes(_indexToFunctionArguments[index].Build().ToArray()));
```

Comment:
This line of code threw a mysterious error once that I haven't been able to reproduce again:

```text
  Message: 
System.ArgumentException : Destination is too short. (Parameter 'destination')

  Stack Trace: 
BuffersExtensions.CopyToMultiSegment[T](ReadOnlySequence`1& sequence, Span`1 destination)
BuffersExtensions.ToArray[T](ReadOnlySequence`1& sequence)
StreamingChatToolCallsBuilder.Build() line 59
ChatExamples.Example04_FunctionCallingStreamingAsync() line 69
ChatExamples.Example04_FunctionCallingStreamingAsync() line 43
GenericAdapter`1.BlockUntilCompleted()
NoMessagePumpStrategy.WaitForCompletion(AwaitAdapter awaiter)
AsyncToSyncAdapter.Await(Func`1 invoke)
TestMethodCommand.RunTestMethod(TestExecutionContext context)
TestMethodCommand.Execute(TestExecutionContext context)
<>c__DisplayClass4_0.<PerformWork>b__0()
<>c__DisplayClass1_0`1.<DoIsolated>b__0(Object _)
ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
ContextUtils.DoIsolated(ContextCallback callback, Object state)
ContextUtils.DoIsolated[T](Func`1 func)
SimpleWorkItem.PerformWork()
```

Reply:
I think I have a bug in the sequence builder code. The line return new ReadOnlySequence<T>(_first, 0, _last!, _last!.Memory.Length); might be wrong. The last parameter to the ctor should be lenght-1, I think

Reply:
Found the issue. The parameters in this constructor are correct. The problem is in `Segment`'s `Append` method in the way that the running index is being calculated.

Instead of using `items.Length` like below:

```csharp
public Segment Append(ReadOnlyMemory<T> items)
{
    long runningIndex;
    checked { runningIndex = RunningIndex + items.Length; }
    Segment segment = new(items, runningIndex);
    Next = segment;
    return segment;
}
```

It should be `Memory.Length` like this:

```csharp
public Segment Append(ReadOnlyMemory<T> items)
{
    long runningIndex;
    checked { runningIndex = RunningIndex + Memory.Length; }
    Segment segment = new(items, runningIndex);
    Next = segment;
    return segment;
}
```

This is because the `RunningIndex` of `Next` node is the `RunningIndex` of the current node plus the length of the current node, whereas `items` here represents the items of the `Next` node.